### PR TITLE
Preserve Anki import order for new cards

### PR DESCRIPTION
**Pull request by Vladyslav-Soldatenko**
## Description

A user reported that ProcrastiLearn shows imported Anki cards in what looks like a random order, even though their source deck (6500 notes) is deliberately sorted by word frequency. This was accurate: new-card selection had no `ORDER BY` at all and picked a random offset, and the `.apkg` parser never read Anki's own new-card position (`cards.due`) — it only read `notes`, discarding the exact signal needed to preserve the deck's order.

This PR adds a persistent ordering column to vocabulary rows, populates it from Anki's `cards.due` on `.apkg` import, and adds a setting to choose between the new sequential order (default) and the previous random order.

## Type of Change

- [x] New feature
- [ ] Bug fix
- [ ] Refactoring (no functional changes)
- [ ] Documentation update
- [ ] Infrastructure / CI/CD
- [ ] Performance improvement

## Changes Made

- Add `position: Long` to `VocabularyEntity` (Room migration 4→5), assigned once at row creation and never modified afterward by reset, edit, or re-import — matches Anki's own "restore original position" default for reset/forgotten cards.
- `AnkiApkgVocabularyParser` now joins `notes`↔`cards` and orders by `cards.due` (Anki's real new-card position), tie-broken by note id for determinism; notes without a `type=0` card sort after every genuinely-new note.
- New "New Card Order" setting (Sequential / Random) in Settings, following the existing `MixMode` pattern end-to-end (DataStore, ViewModel, dialog, settings row). Default is **Sequential**.
- Re-import/duplicate handling rewritten: words are matched by `word` (the actual identity key, case-insensitive) rather than by row `id`. A re-imported word that already exists now updates its translation in place instead of being silently deleted and recreated (which used to wipe FSRS progress). Brand-new rows always get a fresh autogenerated id — including on the JSON restore path — so a coincidental id collision with an unrelated row (e.g. restoring a backup from a different install) can never abort an entire import batch.
- Manual "Add Word" and imports append new rows to the end of the position queue (`MAX(position) + 1`), matching Anki's own behavior for hand-added cards.
- JSON export schema bumped to v4 (`position` field, with a migration step for older exports).

## How to Test

1. Import a `.apkg` deck via Settings → Import and confirm new cards are studied in the deck's own order (frequency order, etc.) rather than randomly.
2. Toggle Settings → New Card Order to "Random" and confirm the previous random behavior returns.
3. Re-import the same or an updated deck and confirm existing progress is preserved, with only genuinely new words appended at the end.
4. Restore a JSON export from Settings → Export/Import and confirm data round-trips.

## Checklist

- [x] Code follows the project's style guidelines
- [x] Self-reviewed the code for obvious errors
- [x] Added or updated tests where applicable, including edge cases
- [x] Existing tests pass locally
- [ ] Updated documentation if needed
- [x] No duplication introduced. I searched the existing code for similar behavior and extracted similar parts to be reused
- [ ] No new warnings or suppressions introduced — **one exception, flagged for review**: `VocabularyDao` now carries a `@Suppress("TooManyFunctions")` (detekt), since it grew past the interface threshold after adding the bulk word-lookup/position/import-batch methods needed for this feature. I judged splitting it into a separate DAO for these CRUD-adjacent methods to be worse than a justified suppression, but flagging per the checklist for maintainer sign-off.

## Related Issues

Addresses a user-reported feature request (no tracked issue number) for controlling new-card order and preserving Anki import order, particularly for large frequency-sorted deck imports.

## Testing performed

Full unit test suite, ktlint, detekt, and Android Lint all pass locally; `assembleDebug` succeeds. Instrumented (`androidTest`) sources compile cleanly but could not be run in this environment (no emulator/device available) — please run the instrumented suite (including the strengthened `SettingsAnkiImportE2eTest`, which now asserts import order end-to-end, and the new `AppDatabaseMigrationTest` migration-4-to-5 cases) in CI or on a device before merging.

---
_Generated by [Claude Code](https://claude.ai/code/session_01CdgrXyxt3QaQG7VWbtJcU1)_

### CODE REVIEW

File: app/src/androidTest/java/com/procrastilearn/app/e2e/SettingsAnkiImportE2eTest.kt [hunk 1]
```diff
@@ -94,6 +95,16 @@ class SettingsAnkiImportE2eTest {
                 actual.isNew,
             )
         }
+
+        // End-to-end proof that the whole pipeline (parser ordering -> merge -> position
+        // assignment) is wired correctly - not just that the right set of words landed.
```
Comment (Vladyslav-Soldatenko) [inline app/src/androidTest/java/com/procrastilearn/app/e2e/SettingsAnkiImportE2eTest.kt]: remove comment

File: app/src/main/java/com/procrastilearn/app/data/export/VocabularyTransferManager.kt [hunk 2]
```diff
@@ -19,11 +20,14 @@ import java.io.File
 import java.io.InputStream
 import javax.inject.Inject
 
+// SQLite's default SQLITE_MAX_VARIABLE_NUMBER is 999 on older Android versions; stay
+// comfortably under it when binding word lists for the duplicate lookup below.
+private const val MAX_SQLITE_BIND_ARGS = 900
+
```
Comment (Vladyslav-Soldatenko) [inline app/src/main/java/com/procrastilearn/app/data/export/VocabularyTransferManager.kt]: don't we already have it somewhere in code? if yes - dont duplicate, extract to common place and reuse

File: app/src/main/java/com/procrastilearn/app/data/export/VocabularyTransferManager.kt [hunk 3]
```diff
@@ -140,17 +144,85 @@ class VocabularyTransferManager
...
             }
 
+        // Imports a parsed batch, merging against existing rows by word (the app's real
+        // identity key - see the unique NOCASE index on VocabularyEntity.word) rather than by
+        // row id, which is per-install and can coincidentally collide across devices/exports.
+        // A word that already exists gets its content fields updated in place, preserving its
+        // id/position/FSRS progress untouched. A genuinely new word is appended to the end of
+        // the position queue and always inserted with id = 0 so Room autogenerates a fresh id -
+        // never carrying over an id from the imported file, which could otherwise coincidentally
+        // collide with an unrelated existing row (see importExportItems).
```
Comment (Vladyslav-Soldatenko) [inline app/src/main/java/com/procrastilearn/app/data/export/VocabularyTransferManager.kt]: remove comment

File: app/src/main/java/com/procrastilearn/app/data/export/VocabularyTransferManager.kt [hunk 4]
```diff
@@ -140,17 +144,85 @@ class VocabularyTransferManager
...
             }
+            vocabularyDao.applyImportBatch(toInsert, toUpdate)
         }
 
+        // Merges a JSON backup/export against existing rows the same way importItems does (by
+        // word, not by id). Matched words update in place, preserving id/position/progress. New
+        // words are inserted with id = 0 (never the id from the export file) so a coincidental
+        // id collision with an unrelated row on this device (e.g. restoring a backup from a
+        // different install) can never abort the whole batch; their exported `position` carries
+        // through verbatim, same as any other export field.
```
Comment (Vladyslav-Soldatenko) [inline app/src/main/java/com/procrastilearn/app/data/export/VocabularyTransferManager.kt]: remove comment

File: app/src/main/java/com/procrastilearn/app/data/export/VocabularyTransferManager.kt [hunk 5]
```diff
@@ -140,17 +144,85 @@ class VocabularyTransferManager
...
+                } else {
+                    toInsert += item.toEntity().copy(id = 0L)
+                }
+            }
+            vocabularyDao.applyImportBatch(toInsert, toUpdate)
+        }
+
+        // Last-occurrence-wins de-dup by case-insensitive word, preserving encounter order for
+        // position assignment. A well-formed deck essentially never has literal duplicate
+        // front-text notes, but this keeps the (rare/adversarial) case well-defined.
```
Comment (Vladyslav-Soldatenko) [inline app/src/main/java/com/procrastilearn/app/data/export/VocabularyTransferManager.kt]: remove comment

File: app/src/main/java/com/procrastilearn/app/data/local/dao/VocabularyDao.kt [hunk 6]
```diff
@@ -5,10 +5,15 @@ import androidx.room.Delete
...
 import androidx.room.OnConflictStrategy
 import androidx.room.Query
+import androidx.room.Transaction
 import androidx.room.Update
 import com.procrastilearn.app.data.local.entity.VocabularyEntity
 import kotlinx.coroutines.flow.Flow
 
+// Arity from covering every read/write shape a single `vocabulary` table CRUD DAO needs
+// (single vs. bulk insert/update/delete, word/id lookups, import support), not from an
+// undecomposed monolith - review/stats-specific queries already live in their own DAOs.
```
Comment (Vladyslav-Soldatenko) [inline app/src/main/java/com/procrastilearn/app/data/local/dao/VocabularyDao.kt]: remove comment

File: app/src/main/java/com/procrastilearn/app/data/local/dao/VocabularyDao.kt [hunk 7]
```diff
@@ -27,15 +32,44 @@ interface VocabularyDao {
...
+    suspend fun getMaxPosition(): Long
+
     @Insert(onConflict = OnConflictStrategy.REPLACE)
     suspend fun insertVocabulary(item: VocabularyEntity): Long
 
-    @Insert(onConflict = OnConflictStrategy.REPLACE)
+    // Deliberately plain ABORT (no REPLACE): callers are expected to have already
+    // split their batch into genuinely-new rows (this) vs existing-row updates
+    // (updateAllVocabulary), via a word-based lookup — see VocabularyTransferManager.
+    // ABORT-on-conflict here is a correctness safety net, not the primary dedup path.
```
Comment (Vladyslav-Soldatenko) [inline app/src/main/java/com/procrastilearn/app/data/local/dao/VocabularyDao.kt]: remove comment

File: app/src/main/java/com/procrastilearn/app/data/local/dao/VocabularyDao.kt [hunk 8]
```diff
@@ -27,15 +32,44 @@ interface VocabularyDao {
...
     suspend fun insertAllVocabulary(items: List<VocabularyEntity>)
 
     @Update
     suspend fun updateVocabulary(item: VocabularyEntity)
 
+    @Update
+    suspend fun updateAllVocabulary(items: List<VocabularyEntity>)
+
+    // Applies a pre-split import batch atomically: brand-new rows are inserted,
+    // rows that matched an existing word (by VocabularyTransferManager) are updated.
```
Comment (Vladyslav-Soldatenko) [inline app/src/main/java/com/procrastilearn/app/data/local/dao/VocabularyDao.kt]: remove comment

File: app/src/main/java/com/procrastilearn/app/data/local/dao/VocabularyReviewDao.kt [hunk 9]
```diff
@@ -115,4 +115,18 @@ interface VocabularyReviewDao {
         offset: Int,
         requireBidirectional: Boolean = false,
     ): Long?
+
+    // Pick the lowest-position "new" row (import/creation order). Tie-broken by id since
+    // position values are not guaranteed unique (e.g. Anki's own cards.due can tie, and
+    // pre-migration-backfilled/default rows can share a position).
```
Comment (Vladyslav-Soldatenko) [inline app/src/main/java/com/procrastilearn/app/data/local/dao/VocabularyReviewDao.kt]: remove comment

File: app/src/main/java/com/procrastilearn/app/data/local/database/Migrations.kt [hunk 10]
```diff
@@ -78,3 +78,19 @@ val MIGRATION_3_4 =
...
     }
+
+val MIGRATION_4_5 =
+    object : Migration(4, 5) {
+        override fun migrate(db: SupportSQLiteDatabase) {
+            db.execSQL("ALTER TABLE `vocabulary` ADD COLUMN `position` INTEGER NOT NULL DEFAULT 0")
+            // SQLite's ADD COLUMN DEFAULT can't express a row-varying value. `id` is already
+            // monotonically increasing in creation order, so reusing it directly preserves
+            // correct relative ordering for existing rows in one O(n) UPDATE, with no
+            // window-function dependency.
```
Comment (Vladyslav-Soldatenko) [inline app/src/main/java/com/procrastilearn/app/data/local/database/Migrations.kt]: remove comment

File: app/src/main/java/com/procrastilearn/app/data/local/entity/VocabularyEntity.kt [hunk 11]
```diff
@@ -13,6 +13,7 @@ import androidx.room.PrimaryKey
         Index(value = ["word"], unique = true),
         Index(value = ["backwardFsrsDueAt"]),
         Index(value = ["fsrsDueAt", "backwardFsrsDueAt"]),
+        Index(value = ["fsrsDueAt", "backwardFsrsDueAt", "position"]),
```
Comment (Vladyslav-Soldatenko) [inline app/src/main/java/com/procrastilearn/app/data/local/entity/VocabularyEntity.kt:16]: explain, why it's needed? what if we need to find only by backwardFsrsDueAt + position?

File: app/src/main/java/com/procrastilearn/app/data/local/entity/VocabularyEntity.kt [hunk 12]
```diff
@@ -36,4 +37,8 @@ data class VocabularyEntity(
     // null = derive backward prompt/answer by swapping translation/word at read time
     val backwardPromptOverride: String? = null,
     val backwardAnswerOverride: String? = null,
+    // New-card queue order, assigned once at creation and never modified afterward
+    // (not by reset, edit, or re-import merge). See VocabularyTransferManager/
+    // VocabularyRepositoryImpl for assignment rules.
```
Comment (Vladyslav-Soldatenko) [inline app/src/main/java/com/procrastilearn/app/data/local/entity/VocabularyEntity.kt]: remove comment

File: app/src/main/java/com/procrastilearn/app/data/parser/anki/AnkiApkgVocabularyParser.kt [hunk 13]
```diff
@@ -21,6 +21,28 @@ private val IMG_TAG_REGEX = Regex("""<img\b[^>]*>""", RegexOption.IGNORE_CASE)
...
+/**
+ * Orders notes by Anki's own new-card position (`cards.due` where `type = 0`), so an imported
+ * deck (e.g. sorted by word frequency) is studied in that same order rather than an arbitrary
+ * one. A note's position is the minimum `due` among its own "new" cards, which also resolves
+ * note types that generate multiple cards per note (e.g. Basic-and-reversed) to a single
+ * deterministic value without special-casing. Notes with no `type = 0` card (already reviewed
+ * in Anki, or no card row at all) sort after every note that has one, tie-broken by `notes.id`
+ * for full determinism (`due` itself is not guaranteed unique). Multiple decks in one file are
+ * intentionally flattened together — nothing here reads or filters by `did`.
+ */
```
Comment (Vladyslav-Soldatenko) [inline app/src/main/java/com/procrastilearn/app/data/parser/anki/AnkiApkgVocabularyParser.kt]: remove comment

File: app/src/main/java/com/procrastilearn/app/data/parser/anki/AnkiApkgVocabularyParser.kt [hunk 14]
```diff
@@ -21,6 +21,28 @@ private val IMG_TAG_REGEX = Regex("""<img\b[^>]*>""", RegexOption.IGNORE_CASE)
...
+    SELECT mid, flds FROM (
+        SELECT n.id AS note_id, n.mid AS mid, n.flds AS flds,
+               MIN(CASE WHEN c.type = 0 THEN c.due END) AS min_new_due
+        FROM notes n
+        LEFT JOIN cards c ON c.nid = n.id
+        GROUP BY n.id
+    )
+    ORDER BY (min_new_due IS NULL) ASC, min_new_due ASC, note_id ASC
+    """.trimIndent()
+
```
Comment (Vladyslav-Soldatenko) [inline app/src/main/java/com/procrastilearn/app/data/parser/anki/AnkiApkgVocabularyParser.kt:35]: what if apkg deck doesnt have order? can it happen?

File: app/src/main/java/com/procrastilearn/app/data/repository/VocabularyRepositoryImpl.kt [hunk 15]
```diff
@@ -81,7 +82,14 @@ class VocabularyRepositoryImpl
                 // Keep FSRS card JSON, but leave dueAt as 0 until first rating.
                 val cardJson = Card.builder().build().toJson()
                 val dueAt = 0L
-                vocabularyDao.insertVocabulary(item.toEntity(fsrsCardJson = cardJson, fsrsDueAt = dueAt))
+                // Read-then-insert must be atomic so two concurrent adds can't read the same
+                // stale max and assign the same position.
```
Comment (Vladyslav-Soldatenko) [inline app/src/main/java/com/procrastilearn/app/data/repository/VocabularyRepositoryImpl.kt]: remove comment

File: app/src/test/java/com/procrastilearn/app/data/export/MigrationsTest.kt [hunk 16]
```diff
@@ -138,4 +146,61 @@ class MigrationsTest {
...
+
+    @Test
+    fun `migrate from 3 runs V3ToV4 and lands exactly on CURRENT_SCHEMA_VERSION`() {
+        val migrated = Migrations.migrate(v3Envelope, from = 3)
+
+        assertThat(migrated.getValue("schemaVersion").jsonPrimitive.int).isEqualTo(CURRENT_SCHEMA_VERSION)
+    }
+
+    @Test
+    fun `V3ToV4 migration decodes a genuine v3 payload without position, defaulting it to zero`() {
```
Comment (Vladyslav-Soldatenko) [inline app/src/test/java/com/procrastilearn/app/data/export/MigrationsTest.kt:174]: why default to 0, not to "id"?

File: app/src/test/java/com/procrastilearn/app/data/parser/anki/AnkiApkgVocabularyParserTest.kt [hunk 17]
```diff
@@ -56,6 +65,102 @@ class AnkiApkgVocabularyParserTest {
...
+                        SyntheticCard(id = 11, nid = 2, type = 0, due = 500),
+                        // type=1 (learning) - also not a new-card position.
+                        SyntheticCard(id = 12, nid = 3, type = 1, due = 2),
+                    ),
+            )
+
+        val result = parser.parse(deckFile)
+
+        // The one genuinely new note comes first; the two non-new notes follow, ordered by
+        // note id since neither has a meaningful due-as-position value.
```
Comment (Vladyslav-Soldatenko) [inline app/src/test/java/com/procrastilearn/app/data/parser/anki/AnkiApkgVocabularyParserTest.kt]: remove comment

File: app/src/test/java/com/procrastilearn/app/data/parser/anki/AnkiApkgVocabularyParserTest.kt [hunk 18]
```diff
@@ -56,6 +65,102 @@ class AnkiApkgVocabularyParserTest {
...
     }
 
+    @Test
+    fun `orders a note with multiple cards by the minimum due among its type=0 cards`() {
+        val deckFile =
+            buildSyntheticApkg(
+                notes =
+                    listOf(
+                        // Two cards per note (e.g. Basic-and-reversed): due 900 and 50.
+                        // The note's position must be the minimum, 50.
```
Comment (Vladyslav-Soldatenko) [inline app/src/test/java/com/procrastilearn/app/data/parser/anki/AnkiApkgVocabularyParserTest.kt]: remove comment

File: app/src/test/java/com/procrastilearn/app/data/parser/anki/AnkiApkgVocabularyParserTest.kt [hunk 19]
```diff
@@ -1,33 +1,42 @@
...
 @RunWith(RobolectricTestRunner::class)
 class AnkiApkgVocabularyParserTest {
     private val parser = AnkiApkgVocabularyParser()
 
     @Test
-    fun `parses vocabulary items from apkg`() {
+    fun `parses vocabulary items from apkg ordered by Anki's own new-card position`() {
+        // Fixture's cards.due (type=0) values: test2=276, TestTitle=8475, bold...=8475
+        // (ties with TestTitle, loses on note id), agree=8476. Order below matches that,
+        // not file/insertion order - see AnkiApkgVocabularyParser's NOTE_ORDER_QUERY.
```
Comment (Vladyslav-Soldatenko) [inline app/src/test/java/com/procrastilearn/app/data/parser/anki/AnkiApkgVocabularyParserTest.kt]: remove comment

File: app/src/test/java/com/procrastilearn/app/ui/screens/settings/SettingsContentTest.kt [hunk 20]
```diff
@@ -102,6 +104,27 @@ class SettingsContentTest {
...
+    @Test
+    fun `selecting new card order option updates callback and closes dialog`() {
+        var selectedOrder: NewCardOrder? = null
+        setContent(onNewCardOrderChange = { selectedOrder = it })
+
+        composeTestRule.onNodeWithText(string(R.string.settings_new_card_order_title)).performClick()
+        composeTestRule.onNodeWithText(string(R.string.settings_new_card_order_random)).performClick()
+
+        assertThat(selectedOrder).isEqualTo(NewCardOrder.RANDOM)
+        // dialog dismissed after selection
```
Comment (Vladyslav-Soldatenko) [inline app/src/test/java/com/procrastilearn/app/ui/screens/settings/SettingsContentTest.kt]: remove comment

File: app/src/androidTest/java/com/procrastilearn/app/e2e/SettingsAnkiImportE2eTest.kt (hunk 1 above)
Comment (Vladyslav-Soldatenko) [reply app/src/androidTest/java/com/procrastilearn/app/e2e/SettingsAnkiImportE2eTest.kt]: Done — comment removed.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: app/src/main/java/com/procrastilearn/app/data/export/VocabularyTransferManager.kt (hunk 2 above)
Comment (Vladyslav-Soldatenko) [reply app/src/main/java/com/procrastilearn/app/data/export/VocabularyTransferManager.kt]: Good catch — `MAX_SQLITE_BIND_ARGS` was already defined in `VocabularyRepositoryImpl`. Extracted it into a shared `data/local/dao/SqliteLimits.kt` and both files now import from there instead of each declaring their own copy. Pushed in 089956e.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: app/src/main/java/com/procrastilearn/app/data/export/VocabularyTransferManager.kt (hunk 3 above)
Comment (Vladyslav-Soldatenko) [reply app/src/main/java/com/procrastilearn/app/data/export/VocabularyTransferManager.kt]: Done — comment removed.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: app/src/main/java/com/procrastilearn/app/data/export/VocabularyTransferManager.kt (hunk 4 above)
Comment (Vladyslav-Soldatenko) [reply app/src/main/java/com/procrastilearn/app/data/export/VocabularyTransferManager.kt]: Done — comment removed.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: app/src/main/java/com/procrastilearn/app/data/export/VocabularyTransferManager.kt (hunk 5 above)
Comment (Vladyslav-Soldatenko) [reply app/src/main/java/com/procrastilearn/app/data/export/VocabularyTransferManager.kt]: Done — comment removed.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: app/src/main/java/com/procrastilearn/app/data/local/dao/VocabularyDao.kt (hunk 6 above)
Comment (Vladyslav-Soldatenko) [reply app/src/main/java/com/procrastilearn/app/data/local/dao/VocabularyDao.kt]: Done — comment removed.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: app/src/main/java/com/procrastilearn/app/data/local/dao/VocabularyDao.kt (hunk 7 above)
Comment (Vladyslav-Soldatenko) [reply app/src/main/java/com/procrastilearn/app/data/local/dao/VocabularyDao.kt]: Done — comment removed.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: app/src/main/java/com/procrastilearn/app/data/local/dao/VocabularyDao.kt (hunk 8 above)
Comment (Vladyslav-Soldatenko) [reply app/src/main/java/com/procrastilearn/app/data/local/dao/VocabularyDao.kt]: Done — comment removed.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: app/src/main/java/com/procrastilearn/app/data/local/dao/VocabularyReviewDao.kt (hunk 9 above)
Comment (Vladyslav-Soldatenko) [reply app/src/main/java/com/procrastilearn/app/data/local/dao/VocabularyReviewDao.kt]: Done — comment removed.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: app/src/main/java/com/procrastilearn/app/data/local/database/Migrations.kt (hunk 10 above)
Comment (Vladyslav-Soldatenko) [reply app/src/main/java/com/procrastilearn/app/data/local/database/Migrations.kt]: Done — comment removed.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: app/src/main/java/com/procrastilearn/app/data/local/entity/VocabularyEntity.kt (hunk 11 above)
Comment (Vladyslav-Soldatenko) [reply app/src/main/java/com/procrastilearn/app/data/local/entity/VocabularyEntity.kt:16]: This composite index exists specifically to serve `pickNewIdByPositionAsc`'s query in `VocabularyReviewDao`:
```sql
WHERE fsrsDueAt = 0 AND backwardFsrsDueAt = 0
ORDER BY position ASC, id ASC
LIMIT 1
```
SQLite can satisfy an equality-prefix + trailing `ORDER BY` from a single index-seek (no separate sort step) only when the equality-filtered columns come first in the index and the sort column immediately follows them — that's why the index is `(fsrsDueAt, backwardFsrsDueAt, position)` in that exact order, matching this query's `WHERE`/`ORDER BY` shape. This is the hot path for picking the next new card to show, so it's worth being index-covered.

If we needed to query by `backwardFsrsDueAt + position` alone (no `fsrsDueAt` predicate), this index wouldn't help — SQLite can only use a *leading-column prefix* of a composite index, and `fsrsDueAt` is the leading column here. No such query exists anywhere in the codebase today: every "new" row lookup filters on both `fsrsDueAt = 0 AND backwardFsrsDueAt = 0` together (that pair is exactly the "new" definition used elsewhere, e.g. `VocabularyStatsDao.countNewTotal`). If a `backwardFsrsDueAt + position`-only query is ever needed, it'd need its own separate index.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: app/src/main/java/com/procrastilearn/app/data/local/entity/VocabularyEntity.kt (hunk 12 above)
Comment (Vladyslav-Soldatenko) [reply app/src/main/java/com/procrastilearn/app/data/local/entity/VocabularyEntity.kt]: Done — comment removed.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: app/src/main/java/com/procrastilearn/app/data/parser/anki/AnkiApkgVocabularyParser.kt (hunk 13 above)
Comment (Vladyslav-Soldatenko) [reply app/src/main/java/com/procrastilearn/app/data/parser/anki/AnkiApkgVocabularyParser.kt]: Done — comment removed.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: app/src/main/java/com/procrastilearn/app/data/parser/anki/AnkiApkgVocabularyParser.kt (hunk 14 above)
Comment (Vladyslav-Soldatenko) [reply app/src/main/java/com/procrastilearn/app/data/parser/anki/AnkiApkgVocabularyParser.kt:35]: Yes, it happens, and the query already handles it (didn't need a code change, but good to confirm it's covered):

1. A note can have no usable "new" due value — either it has zero matching `cards` rows at all (the `LEFT JOIN` gives NULL), or all its cards are already-reviewed in Anki (`type != 0`, so `MIN(CASE WHEN c.type = 0 THEN c.due END)` also evaluates to NULL for that note). Either way `min_new_due` is NULL, and `ORDER BY (min_new_due IS NULL) ASC, min_new_due ASC, note_id ASC` sorts every such note after all notes that do have a real due value, falling back to `note_id` for determinism among themselves.
2. If *every* note in the deck lacks a due value, the parser still returns a valid list — just entirely ordered by note id, no crash.

Added dedicated tests against a synthetic apkg fixture for exactly these cases: `sorts a note with no matching card row after every genuinely new note` and `sorts notes whose only card is not type 0 after every genuinely new note...` in `AnkiApkgVocabularyParserTest`.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: app/src/main/java/com/procrastilearn/app/data/repository/VocabularyRepositoryImpl.kt (hunk 15 above)
Comment (Vladyslav-Soldatenko) [reply app/src/main/java/com/procrastilearn/app/data/repository/VocabularyRepositoryImpl.kt]: Done — comment removed.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: app/src/test/java/com/procrastilearn/app/data/export/MigrationsTest.kt (hunk 16 above)
Comment (Vladyslav-Soldatenko) [reply app/src/test/java/com/procrastilearn/app/data/export/MigrationsTest.kt:174]: This is intentionally different from the SQL Room migration, which does backfill `position` from `id` — the reasoning doesn't carry over cleanly:

For the Room migration, `id` is *this device's own* row-creation order — a genuinely meaningful proxy for "when I added this word." But a v1–v3 JSON export (predating the `position` field) may have been produced on a *different* device. Its `id` values have no reliable relationship to this device's own row ordering — reusing them as `position` could collide with or arbitrarily interleave among positions already assigned locally, which would be actively misleading (a fabricated ordering signal) rather than merely absent.

Defaulting to `0` is the deliberately "safe wrong answer": these legacy items just sort to the very front of the new-card queue — simple, predictable, and it doesn't pretend to know an order it doesn't have. This only affects old exports that predate the field entirely (the JSON restore/backup path), not the primary Anki-`.apkg`-import path this PR is actually about — flagged as an accepted, documented edge case in the migration.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: app/src/test/java/com/procrastilearn/app/data/parser/anki/AnkiApkgVocabularyParserTest.kt (hunk 17 above)
Comment (Vladyslav-Soldatenko) [reply app/src/test/java/com/procrastilearn/app/data/parser/anki/AnkiApkgVocabularyParserTest.kt]: Done — comment removed.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: app/src/test/java/com/procrastilearn/app/data/parser/anki/AnkiApkgVocabularyParserTest.kt (hunk 18 above)
Comment (Vladyslav-Soldatenko) [reply app/src/test/java/com/procrastilearn/app/data/parser/anki/AnkiApkgVocabularyParserTest.kt]: Done — comment removed.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: app/src/test/java/com/procrastilearn/app/data/parser/anki/AnkiApkgVocabularyParserTest.kt (hunk 19 above)
Comment (Vladyslav-Soldatenko) [reply app/src/test/java/com/procrastilearn/app/data/parser/anki/AnkiApkgVocabularyParserTest.kt]: Done — comment removed.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: app/src/test/java/com/procrastilearn/app/ui/screens/settings/SettingsContentTest.kt (hunk 20 above)
Comment (Vladyslav-Soldatenko) [reply app/src/test/java/com/procrastilearn/app/ui/screens/settings/SettingsContentTest.kt]: Done — comment removed.

---
_Generated by [Claude Code](https://claude.ai/code)_
